saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/OpenSamlMetadataResolver.java

```diff
@@ -0,0 +1,160 @@
+
+		SPSSODescriptor spSsoDescriptor = buildObject(builderFactory, SPSSODescriptor.DEFAULT_ELEMENT_NAME);
+		spSsoDescriptor.setAuthnRequestsSigned(registration.getAssertingPartyDetails().getWantAuthnRequestsSigned());
+		spSsoDescriptor.setWantAssertionsSigned(true);
+		spSsoDescriptor.addSupportedProtocol(SAMLConstants.SAML20P_NS);
+
+		NameIDFormat nameIdFormat = buildObject(builderFactory, NameIDFormat.DEFAULT_ELEMENT_NAME);
+		nameIdFormat.setFormat(registration.getAssertingPartyDetails().getNameIdFormat());
+		spSsoDescriptor.getNameIDFormats().add(nameIdFormat);
+
+		spSsoDescriptor.getAssertionConsumerServices().add(
+				buildAssertionConsumerService(registration, builderFactory, request));
+
+		spSsoDescriptor.getKeyDescriptors().addAll(buildKeys(builderFactory,
+				registration.getSigningCredentials(), UsageType.SIGNING));
+		spSsoDescriptor.getKeyDescriptors().addAll(buildKeys(builderFactory,
+				registration.getEncryptionCredentials(), UsageType.ENCRYPTION));
+
+		return spSsoDescriptor;
+	}
+
+	private List<KeyDescriptor> buildKeys(XMLObjectBuilderFactory builderFactory,
+			List<Saml2X509Credential> credentials, UsageType usageType) {
+		List<KeyDescriptor> list = new ArrayList<>();
+		for (Saml2X509Credential credential : credentials) {
+			KeyDescriptor keyDescriptor = buildKeyDescriptor(builderFactory, usageType, credential.getCertificate());
+			list.add(keyDescriptor);
+		}
+		return list;
+	}
+
+	private KeyDescriptor buildKeyDescriptor(XMLObjectBuilderFactory builderFactory, UsageType usageType,
+			java.security.cert.X509Certificate certificate) {
+		KeyDescriptor keyDescriptor = buildObject(builderFactory, KeyDescriptor.DEFAULT_ELEMENT_NAME);
+		KeyInfo keyInfo = buildObject(builderFactory, KeyInfo.DEFAULT_ELEMENT_NAME);
+		X509Certificate x509Certificate = buildObject(builderFactory, X509Certificate.DEFAULT_ELEMENT_NAME);
+		X509Data x509Data = buildObject(builderFactory, X509Data.DEFAULT_ELEMENT_NAME);
+
+		try {
+			x509Certificate.setValue(new String(Base64.getEncoder().encode(certificate.getEncoded())));
+		} catch (CertificateEncodingException e) {
+			throw new Saml2Exception("Cannot encode certificate " + certificate.toString());
+		}
+
+		x509Data.getX509Certificates().add(x509Certificate);
+		keyInfo.getX509Datas().add(x509Data);
+
+		keyDescriptor.setUse(usageType);
+		keyDescriptor.setKeyInfo(keyInfo);
+		return keyDescriptor;
+	}
+
+	private AssertionConsumerService buildAssertionConsumerService(RelyingPartyRegistration registration,
+			XMLObjectBuilderFactory builderFactory, HttpServletRequest request) {
+		AssertionConsumerService assertionConsumerService = buildObject(builderFactory, AssertionConsumerService.DEFAULT_ELEMENT_NAME);
+
+		assertionConsumerService.setLocation(
+				resolveTemplate(registration.getAssertionConsumerServiceLocation(), registration, request));
+		assertionConsumerService.setBinding(registration.getAssertingPartyDetails().getSingleSignOnServiceBinding().getUrn());
```

Comment:
The AssertingPartyDetails#singleSignOnServiceBinding conceptually maps to the `<SingleSignOnService binding='...'/>` attribute.

Instead, I think this should probably be set to `POST` until `RelyingPartyRegistration#assertionConsumerServiceBinding` is added in https://github.com/spring-projects/spring-security/issues/8776

Reply:
You cannot simply hardcode POST here, as REDIRECT is pretty common in Saml. I don't get the problem with such usage as above

Reply:
Are you thinking #8776 should be done first? I think it's okay to wait for that.

> I don't get the problem with such usage as above

The problem with the above usage is that SingleSignOnService#binding is IDP metadata, but you are generating SP metadata. Since you are wanting to show AssertionConsumerService#binding, you'll want to retrieve that from a semantically equivalent source.

> REDIRECT is pretty common in Saml

AFAIK, SAML doesn't allow the usage of REDIRECT for SAML assertions. In [the spec](http://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf) line 421 says:

> In step 5, the identity provider issues a <Response> message to be delivered by the user agent to the service provider. Either the HTTP POST, or HTTP Artifact binding can be used to transfer the message to the service provider through the user agent. The message may indicate an error, or will include (at least) an authentication assertion. The HTTP Redirect binding MUST NOT be used, as the response will typically exceed the URL length permitted by most user agents.

That doesn't mean that ppl don't do it, but it does mean simplifying that specific use case is a lower priority.